### PR TITLE
feat: implementing view/getindex/setindex

### PR DESCRIPTION
**Pull request by avik-pal**

### CODE REVIEW

File: src/TracedRArray.jl [hunk 1]
```diff
@@ -43,6 +43,43 @@ and require expensive copies and synchronization each time and therefore should
...
+
+function Base.setindex!(
+    a::TracedRArray{T,N}, v, indices::Vararg{Union{Base.AbstractUnitRange,Colon},N}
+) where {T,N}
+    indices = [i isa Colon ? (1:size(a, idx)) : i for (idx, i) in enumerate(indices)]
+    @show indices
+    @show v
+    if !(v isa TracedRArray)
+        v = promote_to(TracedRArray{eltype(v),ndims(v)}, v) 
+    end
```
Comment (avik-pal) [inline src/TracedRArray.jl:79]: Am I doing something wrong here? The test code causes a segfault

Comment (wsmoses) [reply src/TracedRArray.jl:79]: it should not segfault? but you can just pass T, N here

Comment (wsmoses) [reply src/TracedRArray.jl:79]: Specifically dynamic slice update takes a tensor of the same type/size so you shouldn't need to check against v

Comment (avik-pal) [reply src/TracedRArray.jl:79]: > it should not segfault? but you can just pass T, N here

<details>
<summary>Details</summary>

```julia
[55793] signal 11 (128): Segmentation fault
in expression starting at REPL[7]:1
_ZN4mlir17DenseElementsAttr3getENS_10ShapedTypeEN4llvm8ArrayRefINS_9AttributeEEE at /mnt/.julia/artifacts/deb4af576455c51b1d426b67767b15a6dd03dfd5/lib/libReactantExtra.so (unknown line)
mlirDenseElementsAttrGet at /mnt/.julia/artifacts/deb4af576455c51b1d426b67767b15a6dd03dfd5/lib/libReactantExtra.so (unknown line)
mlirDenseElementsAttrGet at /mnt/software/lux/Reactant.jl/src/mlir/libMLIR_h.jl:3685 [inlined]
DenseElementsAttribute at /mnt/software/lux/Reactant.jl/src/mlir/IR/Attribute.jl:409
promote_to at /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:176
setindex! at /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:78
unknown function (ip: 0x7a3475111ca2)
update_on_copy at ./REPL[6]:3 [inlined]
opaque closure at ./<missing>:0
unknown function (ip: 0x7a347510eea2)
#22 at /mnt/software/lux/Reactant.jl/src/utils.jl:118
block! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
unknown function (ip: 0x7a347f9d90e6)
#make_mlir_fn#16 at /mnt/software/lux/Reactant.jl/src/utils.jl:74
make_mlir_fn at /mnt/software/lux/Reactant.jl/src/utils.jl:24 [inlined]
#120 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:175 [inlined]
block! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
#119 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:174 [inlined]
mmodule! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Module.jl:93
unknown function (ip: 0x7a347f965ed6)
#compile_to_module!#118 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:171
compile_to_module! at /mnt/software/lux/Reactant.jl/src/Compiler.jl:170 [inlined]
#130 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:336
context! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Context.jl:71
unknown function (ip: 0x7a347f95ed46)
#compile#129 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:334
compile at /mnt/software/lux/Reactant.jl/src/Compiler.jl:329
unknown function (ip: 0x7a347f9543f6)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
do_call at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:126
eval_value at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:223
eval_stmt_value at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:174 [inlined]
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:663
jl_interpret_toplevel_thunk at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:821
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:943
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:625
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:539
jl_interpret_toplevel_thunk at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:821
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:943
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
ijl_toplevel_eval_in at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:994
eval at ./boot.jl:430 [inlined]
eval_user_input at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:226
repl_backend_loop at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:323
#start_repl_backend#59 at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:308
start_repl_backend at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:305
#run_repl#72 at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:464
run_repl at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:450
jfptr_run_repl_11644 at /mnt/.julia/compiled/v1.11/REPL/u0gqU_Db1fB.so (unknown line)
#1137 at ./client.jl:441
jfptr_YY.1137_16324 at /mnt/.julia/compiled/v1.11/REPL/u0gqU_Db1fB.so (unknown line)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
jl_f__call_latest at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/builtins.c:875
#invokelatest#2 at ./essentials.jl:1054 [inlined]
invokelatest at ./essentials.jl:1051 [inlined]
run_main_repl at ./client.jl:425
repl_main at ./client.jl:562 [inlined]
_start at ./client.jl:536
jfptr__start_71892.1 at /home/avikpal/.julia/juliaup/julia-1.11.0-rc3+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
true_main at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/jlapi.c:900
jl_repl_entrypoint at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/jlapi.c:1059
main at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/cli/loader_exe.c:58
unknown function (ip: 0x7a349a565e07)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
Allocations: 31674429 (Pool: 31673646; Big: 783); GC: 34
[1]    55793 segmentation fault (core dumped)  julia --project=. --threads=auto
```

</details>

Comment (avik-pal) [reply src/TracedRArray.jl:79]: same issue persists even if I pass in {T,N}

Comment (wsmoses) [reply src/TracedRArray.jl]: What happens if you set LocalPreferences to use a debug jll?

Comment (avik-pal) [reply src/TracedRArray.jl]: What is the preference for that?

Comment (wsmoses) [reply src/TracedRArray.jl]: mode=dbg iirc cc @mofeing

Comment (wsmoses) [reply src/TracedRArray.jl]: https://github.com/JuliaBinaryWrappers/Reactant_jll.jl/blob/8f1e23ca6909e950d507dc55d9ad99b2a36bc63c/Artifacts.toml#L65

**Comment by wsmoses** [comment]
worst case you can always compile a debug jll by hand with julia --project=deps deps/build_local.jl

File: src/TracedRArray.jl (hunk 1 above)
Comment (avik-pal) [reply src/TracedRArray.jl]: finally got around to trying. even with the debug build I don't seem to get any additional information

Comment (avik-pal) [reply src/TracedRArray.jl]: how about getting a version of this in, that has the view and getindex implemented and hangle setindex later

Comment (wsmoses) [reply src/TracedRArray.jl]: I mean even still down to have setindex in but untested atm and will debug later

**Pull request merged**